src/app-layer-expectation.c

```diff
@@ -0,0 +1,311 @@
+ * \param DFree pointer to function that has to be used to free data
+ *
+ * \return -1 if error
+ * \return 0 if success
+ */
+int AppLayerExpectationCreate(Flow *f, int direction, Port src, Port dst,
+                              AppProto alproto, void *data,
+                              void (* DataFree)(void *))
+{
+    Expectation *iexp = NULL;
+    IPPair *ipp;
+    Address ip_src, ip_dst;
+
+    Expectation *exp = SCCalloc(1, sizeof(*exp));
+    if (exp == NULL)
+        return -1;
+
+    exp->sp = src;
+    exp->dp = dst;
+    exp->alproto = alproto;
+    exp->ts = f->lastts;
+    exp->data = data;
+    exp->direction = direction;
+    exp->DFree = DataFree;
+
+    if (GetFlowAddresses(f, &ip_src, &ip_dst) == -1)
+        goto error;
+    ipp = IPPairGetIPPairFromHash(&ip_src, &ip_dst);
+    if (ipp == NULL)
+        goto error;
+
+    iexp = IPPairGetStorageById(ipp, g_expectation_id);
+    exp->next = iexp;
+    IPPairSetStorageById(ipp, g_expectation_id, exp);
+
+    SC_ATOMIC_ADD(expectation_count, 1);
+    /* As we are creating the expectation, we release lock on IPPair without
+     * setting the ref count to 0. This way the IPPair will be kept till
+     * cleanup */
+    IPPairUnlock(ipp);
+    return 0;
+
+error:
+    if (exp != NULL)
+        SCFree(exp);
+    return -1;
+}
+
+/**
+ * Return Flow storage identifier corresponding to expectation data
+ *
+ * \return expectation data identifier
+ */
+int AppLayerExpectationGetDataId(void)
+{
+    return g_expectation_data_id;
+}
+
+static Expectation * RemoveExpectationAndGetNext(IPPair *ipp,
+                                Expectation *pexp, Expectation *exp,
+                                Expectation *lexp)
+{
+    /* we remove the object so we get ref count down by 1 to remove reference
+     * hold by the expectation
+     */
+    (void) IPPairDecrUsecnt(ipp);
+    SC_ATOMIC_SUB(expectation_count, 1);
+    if (pexp == NULL) {
+        IPPairSetStorageById(ipp, g_expectation_id, lexp);
+    } else {
+        pexp->next = lexp;
+    }
+    if (exp->data)
+        SCFree(exp->data);
+    SCFree(exp);
+    return lexp;
+}
+
+/**
+ * Function doing a lookup in expectation list and updating Flow if needed.
+ *
+ * This function lookup for a existing expectation that could match the Flow.
+ * If found and if the expectation contains data it store the data in the
+ * expectation storage of the Flow.
+ *
+ * \return an AppProto value if found
+ * \return ALPROTO_UNKNOWN if not found
+ */
+AppProto AppLayerExpectationHandle(Flow *f, int direction)
+{
+    AppProto alproto = ALPROTO_UNKNOWN;
+    IPPair *ipp = NULL;
+    Expectation *lexp = NULL;
+    Expectation *pexp = NULL;
+
+    int x = SC_ATOMIC_GET(expectation_count);
+    if (x == 0) {
+        return ALPROTO_UNKNOWN;
+    }
+
+    Expectation *exp = AppLayerExpectationGet(f, direction, &ipp);
+    time_t ctime = f->lastts.tv_sec;
+
+    if (exp == NULL)
+        goto out;
+
+    pexp = NULL;
+    while (exp) {
```

Comment:
it seems that if we have multiple expectations that apply to a flow each of them would be used to update the flow. Would FlowSetStorageById handle that case w/o loosing memory?